File: bent/fermat2.lua
-- Lua version of the Fermat problem solution
-- Improved from fermat.lua

dofile("primes.lua")

dumpfactor = function(x)
    for i,j in pairs(x) do
        print(j[1],j[2])
    end
end


sumfactors = function(x)
    s=1
    for i,j in pairs(x) do
        t=1
        for k=1,j[2] do
            t=t*j[1]+1
        end
        s=s*t
    end
    return(s)
end

-- Search for numbers that have an interesting property noted by
-- Fermat.  The sum of the factors of the square of a number are
-- a perfect cube.  The first of these numbers is 1.  What is the
-- next larger integer with this property?

search_solution = function(n1,n2)
    local eps
    local x
    local c
    local n

    eps=1e-6

    n1=n1 or 1
    n2=n2 or 100

    for n=n1,n2 do
        x=factorint(n)
        for k=1,#x do x[k][2]=2*x[k][2] end
        target=sumfactors(x)
        c=math.floor(math.pow(target,1/3)+eps)
        if (c*c*c == target) then
            print("Solution = " .. n)
        end
    end
end

File: bent/fermat2.py
import sympy
import time

## From Bent: 2008_2 Computer Bonus
## Computer Bonus
## 
## Find the smallest integer greater than 7 which has the property that
## the sum of all the integral divisors of its cube equals a perfect
## square. One and N are both divisors of N.
## 
## -- Pierre de Fermat (circa 1650)
##
##

max_n = 1_000_000
solutions = []

t1=time.time()
for n in range(8,max_n+1):
    x=sympy.factorint(n*n*n)
    s=1
    for k in x:
        t=1
        for m in range(x[k]):
            t=k*t+1
        s=s*t
    # s = sum of the factors of the cube of n
    # Check to see if it is a perfect square
    c=round(s**(1/2))
    if c*c == s:
        print("Solution = ",n,"   Time = ",time.time()-t1)
        solutions.append(n)
t2=time.time()
print("Total elapsed time = ",t2-t1)
print("Largest value tested = ",max_n)
print("Number of solutions = ",len(solutions))
print("Solution set:")
print(solutions)

